Exclude parent operator time in GpuShuffledSymmetricHashJoin build time

thirtiseven opened the pull request:
Closes https://github.com/NVIDIA/spark-rapids/issues/12529

This pr fixes `build time` metric in `GpuShuffledSymmetricHashJoin` to exclude parent operator time, and only included the real time to build.

Before this pr:
![Screenshot 2025-04-18 at 15 23 35](https://github.com/user-attachments/assets/d79206f8-17de-40d2-9bb3-3600e68c6b91)



After this pr:

![Screenshot 2025-04-18 at 15 15 08](https://github.com/user-attachments/assets/b648743d-c32e-44a4-9cf9-b1ecf1670392)


<!--

Thank you for contributing to RAPIDS Accelerator for Apache Spark!

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present).

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please avoid
   rebasing your branch during the review process, as this causes the context
   of any comments made by reviewers to be lost. If conflicts occur during
   review then they should be resolved by merging into the branch used for
   making the pull request.

Many thanks in advance for your cooperation!

-->

binmahone commented:
after your PR the metrics seems all broke, double check? 
<img width="463" alt="image" src="https://github.com/user-attachments/assets/95b2502c-32fe-46b7-a347-8f6e0649fa8d" />

thirtiseven commented:
> after your PR the metrics seems all broke, double check?

Thanks, I mistakenly removed the build size metric, and the build time in my test running to fast to show a number. Fixed and ran a different experiment, updated in description.

binmahone commented:
Hi @thirtiseven , I'm afraid we're not working on the right direction for this PR. a few points:

1. It seems that buildTime and streamTime are designed to capture how much time it took to fetch input from build/stream side, so by nature it should contain parent operator time. The problem is that the term "build time" is ambiguous, maybe better called "fetchFromBuildSideTime" and "fetchFromStreamSideTime"
2. The real issue is that today OP TIME for join will contain parent operator time. I'm thinking that we can get the real OP TIME by minus fetchFromBuildSideTime and fetchFromStreamSideTime by the time join finishes, given that every step in join is sequential (need to confirm)
3. I saw your PR only covered GpuShufffledSizedHashJoin, but GpuBroadcastHashjJoin and GpuShuffledHashJoin also need to change
4. I suggest to test the whole PR in some real workloads so that we can tell if the numbers are reasonable. The example you paste is just to simple to see if everything is as expected.

cc @winningsix

thirtiseven commented:
Offline synced with @binmahone, we think the current `build time` / `stream time` metric is acceptable and `op time` does not include parent operator time. And there is no *real build time* in the `GpuShuffledSymmetricHashJoin` that will consume much time.

> It seems that buildTime and streamTime are designed to capture how much time it took to fetch input from build/stream side, so by nature it should contain parent operator time. The problem is that the term "build time" is ambiguous, maybe better called "fetchFromBuildSideTime" and "fetchFromStreamSideTime"

As Mahone said, in `GpuShuffledSymmetricHashJoin`, the build time and stream time are more like fetchFromBuildSideTime and fetchFromStreamSideTime. And actually there is no real `build time` after fetching the ColumnarBatch, like building a hash table or some time consuming preprocessing. The `baseBuildIter` will just concat the batches and send them to join.

Now the `build time` and `stream time` can also match the time in their parent operator, and no other time seems to be spent:

![Screenshot 2025-04-29 at 18 00 07](https://github.com/user-attachments/assets/77317a20-ee73-415e-b709-b09b027afb82)



> The real issue is that today OP TIME for join will contain parent operator time. I'm thinking that we can get the real OP TIME by minus fetchFromBuildSideTime and fetchFromStreamSideTime by the time join finishes, given that every step in join is sequential (need to confirm)

Also, the op time in `GpuShuffledSymmetricHashJoin` excluded parent operator time already.

https://github.com/NVIDIA/spark-rapids/blob/f8d969d89a7489521a735a595396cfb9eb50a2dd/sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala#L98-L105

The `setupNextGatherer` will introduce the parent operator time, but has been excluded from the join op time.

> I saw your PR only covered GpuShufffledSizedHashJoin, but GpuBroadcastHashjJoin and GpuShuffledHashJoin also need to change

I checked the build time metric for other joins, they looks fine to me. Got some screenshots:

GpuBroadcastHashJoin does not have build time metric
<img width="590" alt="Screenshot 2025-04-29 at 17 01 17" src="https://github.com/user-attachments/assets/cb2c28a0-1944-4a05-afd4-9a739a9d385c" />


GpuBroadcastHashjJoin also does not have build time metric
<img width="564" alt="Screenshot 2025-04-29 at 17 30 56" src="https://github.com/user-attachments/assets/cc63fa4d-f8f1-4ae5-92dd-896c64b18476" />

GpuShuffledHashJoin's build time is the real build time and not include the parent operator time.

<img width="530" alt="Screenshot 2025-04-29 at 17 42 06" src="https://github.com/user-attachments/assets/3d1da736-3fbc-4455-ba6c-89079bb2e2e8" />


So I suggest we close this pr and the issue. What do you think @winningsix

thirtiseven commented:
Close this because the current metrics look fine
